static/js/message_list_view.js

```diff
@@ -1030,60 +1011,17 @@ export class MessageListView {
-        const scroll_limit = this._scroll_limit($selected_row, info);
-
-        // This next decision is fairly debatable.  For a big message that
-        // would push the pointer off the screen, we do a partial autoscroll,
-        // which has the following implications:
-        //    a) user sees scrolling (good)
-        //    b) user's pointer stays on screen (good)
-        //    c) scroll amount isn't really tied to size of new messages (bad)
-        //    d) all the bad things about scrolling for users who want messages
-        //       to stay on the screen
-        let scroll_amount;
-        let need_user_to_scroll;
-
-        if (new_messages_height <= scroll_limit) {
-            // This is the happy path where we can just scroll
-            // automatically, and the user will see the new message.
-            scroll_amount = new_messages_height;
-            need_user_to_scroll = false;
-        } else {
-            // Sometimes we don't want to scroll the entire height of
-            // the message, but our callers can give appropriate
-            // warnings if the message is gonna be offscreen.
-            // (Even if we are somewhat constrained here, the message may
-            // still end up being visible, so we do some arithmetic.)
-            scroll_amount = scroll_limit;
-            const offset = $message_viewport.offset_from_bottom($last_visible);
-
-            // For determining whether we need to show the user a "you
-            // need to scroll down" notification, the obvious check
-            // would be `offset > scroll_amount`, and that is indeed
-            // correct with a 1-line message in the compose box.
-            // However, the compose box is open with the content of
-            // the message just sent when this code runs, and
-            // `offset_from_bottom` if an offset from the top of the
-            // compose box, which is about to be reset to empty.  So
-            // to compute the offset at the time the user might see
-            // this notification, we need to adjust by the amount that
-            // the current compose is bigger than the empty, open
-            // compose box.
-            const compose_textarea_default_height = 42;
-            const compose_textarea_current_height = $("#compose-textarea").height();
-            const expected_change =
-                compose_textarea_current_height - compose_textarea_default_height;
-            const expected_offset = offset - expected_change;
-            need_user_to_scroll = expected_offset > scroll_amount;
-        }
-
-        // Ok, we are finally ready to actually scroll.
+        // Find out how close we are to the bottom of the feed. The +1 is to make sure we get
+        // consistent behavior when composing a message right after an autoscroll.
+        const scroll_amount = $message_viewport.offset_from_bottom($last_visible, true) + 1;
         if (scroll_amount > 0) {
```

Comment:
Will this always be true? Or can  `$message_viewport.offset_from_bottom` be negative?

Reply:
As specified in $message_viewport.offset_from_bottom's comment, the return value is positive if the last row is hidden. If we scroll all the way to the bottom past the last row, the return value would be negative.

Reply:
OK, so this is a check on whether the previous last message was visible or not?

Also, why is that the correct amount to scroll, then? Our previous logic based the scroll amount on the height of the new message; what's the thinking behind this calculation being correct?

Reply:
So the previous behavior had a case where it did a partial scroll if the height of the new message was smaller than the scroll_limit, with the scroll_limit being the distance between the selected message and the top of the screen.

Since we no longer care about pushing the selected message off screen, the logic can be simplified. This calculation is sufficient since the height of the new message is included in the offset_from_bottom calculation. As long as the bottom of the last message will be visible when the compose box is empty, we want to scroll to the bottom of the feed.